File: software/mobile/android/app-mobile/src/main/kotlin/com/avlyubimov/svc/mobile/RideDesignTokens.kt
package com.avlyubimov.svc.mobile

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.avlyubimov.svc.core.model.RideResolvedTheme

internal data class RidePalette(
    val background: Color,
    val surface: Color,
    val raisedSurface: Color,
    val primaryText: Color,
    val secondaryText: Color,
    val divider: Color,
    val accent: Color,
    val accentBright: Color,
    val warning: Color,
    val critical: Color,
    val valid: Color,
    val degraded: Color,
) {
    companion object {
        fun resolve(theme: RideResolvedTheme): RidePalette = when (theme) {
            RideResolvedTheme.DAY -> RidePalette(
                background = Color(0xFF0B1118),
                surface = Color(0xFF141D27),
                raisedSurface = Color(0xFF1C2733),
                primaryText = Color.White,
                secondaryText = Color(0xFFBEC9D4),
                divider = Color(0xFF404E5C),
                accent = Color(0xFF1C69D4),
                accentBright = Color(0xFF4AA3FF),
                warning = Color(0xFFFFB000),
                critical = Color(0xFFE32636),
                valid = Color(0xFF5BD690),
                degraded = Color(0xFFFFB000),
            )
            RideResolvedTheme.NIGHT -> RidePalette(
                background = Color(0xFF030508),
                surface = Color(0xFF0A0E13),
                raisedSurface = Color(0xFF11171E),
                primaryText = Color(0xFFF4F7FA),
                secondaryText = Color(0xFF97A5B3),
                divider = Color(0xFF28323D),
                accent = Color(0xFF1454B1),
                accentBright = Color(0xFF398BE8),
                warning = Color(0xFFFFB000),
                critical = Color(0xFFE32636),
                valid = Color(0xFF4AC27E),
                degraded = Color(0xFFFFB000),
            )
        }
    }
}

internal object RideDesignTokens {
    val cornerRadius = 18.dp
    val compactCornerRadius = 12.dp
    val spacing = 14.dp
    val compactSpacing = 8.dp
    val contentPadding = 18.dp
    val statusHeight = 34.dp
    val gaugeLineWidth = 16.dp
    val metricMinimumWidth = 128.dp
    const val standardAnimationMillis = 180
    const val themeAnimationMillis = 300
}

File: software/mobile/ios/SVCMobile/Sources/Model/RideDesignTokens.swift
import SwiftUI

enum RideThemeMode: String, CaseIterable, Identifiable {
    case day
    case night
    case automatic

    var id: String { rawValue }

    var title: String {
        switch self {
        case .day: "SVC Day"
        case .night: "SVC Night"
        case .automatic: "Automatic"
        }
    }
}

enum RideResolvedTheme: Equatable {
    case day
    case night
}

struct RideThemeThresholds: Equatable {
    let nightEnterLux: Double
    let dayEnterLux: Double

    static let `default` = RideThemeThresholds(
        nightEnterLux: 250,
        dayEnterLux: 650
    )
}

enum RideThemeResolver {
    static func resolve(
        mode: RideThemeMode,
        ambientLight: RideValue<Double>,
        previous: RideResolvedTheme
    ) -> RideResolvedTheme {
        resolve(
            mode: mode,
            ambientLight: ambientLight,
            previous: previous,
            thresholds: .default
        )
    }

    static func resolve(
        mode: RideThemeMode,
        ambientLight: RideValue<Double>,
        previous: RideResolvedTheme,
        thresholds: RideThemeThresholds
    ) -> RideResolvedTheme {
        switch mode {
        case .day:
            return .day
        case .night:
            return .night
        case .automatic:
            guard let lux = ambientLight.displayValue else {
                return previous
            }
            switch previous {
            case .day:
                return lux <= thresholds.nightEnterLux ? .night : .day
            case .night:
                return lux >= thresholds.dayEnterLux ? .day : .night
            }
        }
    }
}

struct RidePalette {
    let background: Color
    let surface: Color
    let raisedSurface: Color
    let primaryText: Color
    let secondaryText: Color
    let divider: Color
    let accent: Color
    let accentBright: Color
    let warning: Color
    let critical: Color
    let valid: Color
    let degraded: Color

    static func resolve(_ theme: RideResolvedTheme) -> RidePalette {
        switch theme {
        case .day:
            RidePalette(
                background: Color(red: 11 / 255, green: 17 / 255, blue: 24 / 255),
                surface: Color(red: 20 / 255, green: 29 / 255, blue: 39 / 255),
                raisedSurface: Color(red: 28 / 255, green: 39 / 255, blue: 51 / 255),
                primaryText: .white,
                secondaryText: Color(red: 190 / 255, green: 201 / 255, blue: 212 / 255),
                divider: Color(red: 64 / 255, green: 78 / 255, blue: 92 / 255),
                accent: Color(red: 28 / 255, green: 105 / 255, blue: 212 / 255),
                accentBright: Color(red: 74 / 255, green: 163 / 255, blue: 1),
                warning: Color(red: 1, green: 176 / 255, blue: 0),
                critical: Color(red: 227 / 255, green: 38 / 255, blue: 54 / 255),
                valid: Color(red: 91 / 255, green: 214 / 255, blue: 144 / 255),
                degraded: Color(red: 1, green: 176 / 255, blue: 0)
            )
        case .night:
            RidePalette(
                background: Color(red: 3 / 255, green: 5 / 255, blue: 8 / 255),
                surface: Color(red: 10 / 255, green: 14 / 255, blue: 19 / 255),
                raisedSurface: Color(red: 17 / 255, green: 23 / 255, blue: 30 / 255),
                primaryText: Color(red: 244 / 255, green: 247 / 255, blue: 250 / 255),
                secondaryText: Color(red: 151 / 255, green: 165 / 255, blue: 179 / 255),
                divider: Color(red: 40 / 255, green: 50 / 255, blue: 61 / 255),
                accent: Color(red: 20 / 255, green: 84 / 255, blue: 177 / 255),
                accentBright: Color(red: 57 / 255, green: 139 / 255, blue: 232 / 255),
                warning: Color(red: 1, green: 176 / 255, blue: 0),
                critical: Color(red: 227 / 255, green: 38 / 255, blue: 54 / 255),
                valid: Color(red: 74 / 255, green: 194 / 255, blue: 126 / 255),
                degraded: Color(red: 1, green: 176 / 255, blue: 0)
            )
        }
    }
}

enum RideDesignTokens {
    static let cornerRadius: CGFloat = 18
    static let compactCornerRadius: CGFloat = 12
    static let spacing: CGFloat = 14
    static let compactSpacing: CGFloat = 8
    static let contentPadding: CGFloat = 18
    static let statusHeight: CGFloat = 34
    static let gaugeLineWidth: CGFloat = 16
    static let metricMinimumWidth: CGFloat = 128
    static let standardAnimation = 0.18
    static let themeAnimation = 0.3
}

enum RideMotionPolicy {
    static func duration(
        reduceMotion: Bool,
        standardDuration: Double = RideDesignTokens.standardAnimation
    ) -> Double {
        reduceMotion ? 0 : standardDuration
    }
}
